src/if_statements.py

```diff
@@ -570,6 +603,216 @@ def try_make_if_condition(
+            continue
+
+        elif start not in node.dominators or end not in node.postdominators:
+            # The node must be within the [start, end] interval
+            pass
+
+        elif isinstance(node, BasicNode):
+            # This node has no statements, so it is just a goto
+            nodes_to_mark_emitted.add(node)
+            node_queue.append(node.successor)
+            continue
+
+        elif isinstance(node, ConditionalNode):
+            # If this is a "switch guard" `if` statement, continue iterating on both branches
+            control_expr = switch_guard_expr(node)
+            if (
+                control_expr is not None
+                and early_unwrap_ints(control_expr) == uw_var_expr
+            ):
+                node_queue.append(node.fallthrough_edge)
+                node_queue.append(node.conditional_edge)
+                nodes_to_mark_emitted.add(node)
+                continue
+
+            # Check if the branch_condition is a comparison between var_expr and a Literal
+            assert block_info.branch_condition is not None
+            cond = simplify_condition(block_info.branch_condition)
+            if (
+                isinstance(cond, BinaryOp)
+                and isinstance(cond.right, Literal)
+                and early_unwrap_ints(cond.left) == uw_var_expr
+            ):
+                # IDO typically uses `x == N` and `x < N` patterns in these if trees, but it
+                # will use `x != N` when it needs to jump backwards to an already-emitted block.
+                # GCC will freely use other types of comparisons
+                # Examples from PM: func_8026E558, pr_load_npc_extra_anims
+                if cond.op == "==":
+                    if cond.right.value in cases:
+                        return None
+                    cases[cond.right.value] = node.conditional_edge
+                    node_queue.append(node.fallthrough_edge)
+                elif cond.op == "!=" and (
+                    node.block.index > node.conditional_edge.block.index
+                    or context.options.compiler == context.options.CompilerEnum.GCC
+                ):
+                    if cond.right.value in cases:
+                        return None
+                    cases[cond.right.value] = node.fallthrough_edge
+                    node_queue.append(node.conditional_edge)
+                elif cond.op == "<" or (
+                    cond.op in (">=", "<=", ">")
+                    and context.options.compiler == context.options.CompilerEnum.GCC
+                ):
+                    # TODO: GCC (maybe?) can emit a series of `<`/`>=` comparisons that limit
+                    # to a specific case. Ex: `x < 9 && x >= 8` is only true for `x == 8`.
+                    # Detecting these would require tracking & checking all of the bounds.
+                    # For now, ignore those, and assume the compiler isn't emitting nonsense.
```

Comment:
(if it emits e.g. `8 <= x < 16` to cover case 8..15: I wouldn't call that nonsense)

Reply:
Ah yeah I should clarify. By "nonsense," I meant something structured like this:
```c
if (x > 10) {
    if (x == 5) {
        foo();
    } else if (x == 15) {
        bar();
    }
} else {
    if (x == 5) {
        baz();
    }
}
```
This would be equivalent to the following, because `foo()` is unreachable. But the current logic would not make a switch statement, because it looks like `x == 5` is repeated.
```c
switch (x) {
    case 5:  baz(); break;  // the `foo();` branch was unreachable
    case 15: bar(); break;
}
```